src/mpi-base.jl

```diff
@@ -5,11 +5,73 @@ typealias MPIDatatype Union{Char,
 typealias MPIBuffertype{T} Union{Ptr{T}, Array{T}, Ref{T}}
 
 # Define a function mpitype(T) that returns the MPI datatype code for
-# a given type T. The dictonary is defined in __init__ so the module
-# can be precompiled
+# a given type T. In the case the the type does not exist, it is created and
+# then returned. The dictonary is defined in __init__ so the module can be
+# precompiled
+
+# accessor and creation function for getting MPI datatypes
+function mpitype{T}(::Type{T})
+
+    if haskey(mpitype_dict, T)  # if the datatype already exists
```

Comment:
This `haskey` check is not necessary since the first thing that `Type_create_struct` does is check the dict. You could merge the two functions into one.